sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubIOJsonTable.java

```diff
@@ -123,12 +141,32 @@ public BeamIOType getSourceType() {
 
   @Override
   public PCollection<Row> buildIOReader(Pipeline pipeline) {
-    return
+    PCollectionTuple rowsWithDlq =
         PBegin
             .in(pipeline)
             .apply("readFromPubsub", readMessagesWithAttributes())
-            .apply("parseMessageToRow", PubsubMessageToRow.forSchema(getSchema()))
-            .setCoder(getSchema().getRowCoder());
+            .apply("parseMessageToRow", createParserParDo());
```

Comment:
A note for the future is that building the layers very carefully as composite PTansforms gives good rendering in UIs and places you can add things like DisplayData.

So you might have the nesting:

 - PubsubTable
   - PubsubIO
   - ParseToRow
      - JsonToRowWithDeadLetter
      - WriteDeadLetter